### PR TITLE
Add debian support

### PR DESCRIPTION
**Pull request by wbxyz**

### CODE REVIEW

File: pkg/rebuild/rebuild/registry.go [hunk 1]
```diff
@@ -67,6 +76,13 @@ func warmCacheforArtifact(ctx context.Context, registry RegistryMux, t Target) {
 		registry.CratesIO.Crate(ctx, t.Package)
 		registry.CratesIO.Version(ctx, t.Package, t.Version)
 		registry.CratesIO.Artifact(ctx, t.Package, t.Version)
+	case Debian:
+		component, name, found := strings.Cut(t.Package, "/")
+		if !found {
+			log.Printf("warming cache failed, expected component in debian package name %s", t.Package)
+		}
```
Comment (msuozzo) [inline pkg/rebuild/rebuild/registry.go:83]: the warmups should be in an else or, alternatively, you could early return in the !found

File: pkg/registry/debian/debian.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	}
+	if resp.StatusCode != 200 {
+		return nil, errors.Errorf("fetching artifact: %v", resp.Status)
+	}
+	return resp.Body, nil
+}
+
+func PoolURL(component, name, artifact string) string {
+	prefixDir := name[0:1]
+	// Most packages are in a prefix dir matching their first letter. The exception are lib* files which are in dirs matching lib*.
```
Comment (msuozzo) [inline pkg/registry/debian/debian.go:68]: nit: Maybe just kill the first sentence. You mainly want to describe the exceptional case: "When packages begin with "lib," the dir prefix excludes that portion of the name."

**Review by msuozzo — COMMENTED** [review]
looks good! just a few more tests and I think we're there

File: pkg/registry/debian/debian.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+			// Handle new field.
+			field, value, found := strings.Cut(line, ":")
+			if !found {
+				return nil, fmt.Errorf("expected new field: %v", line)
+			}
+			if _, ok := stanza.Fields[field]; ok {
+				return nil, fmt.Errorf("duplicate field in stanza: %s", field)
+			}
+			stanza.Fields[field] = []string{}
+			// Skip empty first lines (start of a multiline field).
```
Comment (msuozzo) [inline pkg/registry/debian/debian.go:119]: nit: "multiline or folded field"

File: pkg/registry/debian/debian.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+		} else if strings.HasPrefix(line, " ") || strings.HasPrefix(line, "\t") {
+			// Handle continuation lines.
+			if lastField != "" {
+				stanza.Fields[lastField] = append(stanza.Fields[lastField], strings.TrimSpace(line))
+			}
+		} else {
+			// Handle new field.
+			field, value, found := strings.Cut(line, ":")
+			if !found {
+				return nil, fmt.Errorf("expected new field: %v", line)
```
Comment (msuozzo) [inline pkg/registry/debian/debian.go:113]: nit: errors.Errorf

File: pkg/registry/debian/debian.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+		// Check for PGP signature footer.
+		if strings.HasPrefix(b.Text(), "-----BEGIN PGP SIGNATURE-----") {
+			break
+		}
+		line := b.Text()
+		if strings.TrimSpace(line) == "" {
+			// Handle empty lines as stanza separators.
+			if len(stanza.Fields) > 0 {
+				d.Stanzas = append(d.Stanzas, stanza)
+				stanza = ControlStanza{Fields: map[string][]string{}}
```
Comment (msuozzo) [inline pkg/registry/debian/debian.go:102]: i think lastField should also be reset here.

File: pkg/registry/debian/debian.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+			}
+		} else if strings.HasPrefix(line, " ") || strings.HasPrefix(line, "\t") {
+			// Handle continuation lines.
+			if lastField != "" {
+				stanza.Fields[lastField] = append(stanza.Fields[lastField], strings.TrimSpace(line))
+			}
```
Comment (msuozzo) [inline pkg/registry/debian/debian.go:108]: i think the else case should return an error

File: pkg/rebuild/debian/infer.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	}
+	dscURI, dsc, err := mux.Debian.DSC(ctx, component, name, t.Version)
+	if err != nil {
+		return nil, err
+	}
+	var orig, debianTar FileWithChecksum
+	var dependencies []string
+	for stanza := range dsc.Stanzas {
+		for field, values := range dsc.Stanzas[stanza].Fields {
+			if field == "Files" {
```
Comment (msuozzo) [inline pkg/rebuild/debian/infer.go:49]: change this to a switch..case?

File: pkg/rebuild/debian/infer.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+					if strings.HasSuffix(f, ".orig.tar.xz") {
+						orig.URL = debreg.PoolURL(component, name, f)
+						orig.MD5 = md5
+						continue
+					} else if strings.HasSuffix(f, ".debian.tar.xz") {
+						debianTar.URL = debreg.PoolURL(component, name, f)
+						debianTar.MD5 = md5
+						continue
+					}
+					// TODO: support natvie debian packages (single .tar.xz file)
```
Comment (msuozzo) [inline pkg/rebuild/debian/infer.go:66]: nit: spelling

File: pkg/rebuild/debian/infer.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+	"strings"
+
+	billy "github.com/go-git/go-billy/v5"
+	"github.com/go-git/go-git/v5/storage"
+	"github.com/google/oss-rebuild/pkg/rebuild/rebuild"
+	debreg "github.com/google/oss-rebuild/pkg/registry/debian"
+	"github.com/pkg/errors"
+)
+
+func (Rebuilder) InferRepo(_ context.Context, _ rebuild.Target, _ rebuild.RegistryMux) (string, error) {
```
Comment (msuozzo) [inline pkg/rebuild/debian/infer.go:28]: want to add quick comments to these about why they're no-ops?

File: pkg/rebuild/debian/rebuild.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+			return nil, errors.Wrapf(err, "failed to read upstream artifact")
+		}
+	}
+
+	if rbb.Len() > upb.Len() {
+		return errors.New("rebuild is larger than upstream"), nil
+	} else if rbb.Len() < upb.Len() {
+		return errors.New("upstream is larger than rebuild"), nil
+	}
+	upbytes := upb.Bytes()
```
Comment (msuozzo) [inline pkg/rebuild/debian/rebuild.go]: bytes.Equal()

File: pkg/rebuild/debian/strategy.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+wget {{.Debian.URL}}
+dpkg-source -x --no-check $(basename "{{.DSC.URL}}")
+	`, b)
+	if err != nil {
+		return rebuild.Instructions{}, err
+	}
+	deps, err := rebuild.PopulateTemplate(`
+set -eux
+apt update
+{{range .Requirements -}}
```
Comment (msuozzo) [inline pkg/rebuild/debian/strategy.go:54]: nit: you should be able to collapse these into a single `apt install`

File: pkg/rebuild/debian/strategy.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+		Target   rebuild.Target
+	}{*b, be, t})
+	if err != nil {
+		return rebuild.Instructions{}, err
+	}
+	build, err := rebuild.PopulateTemplate(`
+set -eux
+cd */
+debuild -b -uc -us && 
+pwd
```
Comment (msuozzo) [inline pkg/rebuild/debian/strategy.go:69]: remove these

File: pkg/rebuild/debian/strategy.go [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+		BuildEnv rebuild.BuildEnv
+		Target   rebuild.Target
+	}{*b, be, t})
+	if err != nil {
+		return rebuild.Instructions{}, err
+	}
+	build, err := rebuild.PopulateTemplate(`
+set -eux
+cd */
+debuild -b -uc -us &&
```
Comment (msuozzo) [inline pkg/rebuild/debian/strategy.go:68]: remove the trailing &&

File: pkg/rebuild/debian/strategy.go [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+	if err != nil {
+		return rebuild.Instructions{}, err
+	}
+	return rebuild.Instructions{
+		Location:   rebuild.Location{},
+		Source:     src,
+		Deps:       deps,
+		Build:      build,
+		SystemDeps: []string{"wget", "git", "build-essential", "fakeroot", "devscripts"},
+		OutputPath: path.Join(".", t.Artifact),
```
Comment (msuozzo) [inline pkg/rebuild/debian/strategy.go:82]: remove the ".". just the artifact name should be sufficient, right?

File: pkg/rebuild/rebuild/rebuildremote.go [hunk 15]
```diff
@@ -342,11 +383,20 @@ func makeDockerfile(input Input, opts RemoteOptions) (string, error) {
...
-		UtilPrebuildBucket: opts.UtilPrebuildBucket,
-		Instructions:       instructions,
-	})
+	if input.Target.Ecosystem == Debian {
+		err = debuildContainerTpl.Execute(dockerfile, rebuildContainerArgs{
+			UseTimewarp:        opts.UseTimewarp,
+			UtilPrebuildBucket: opts.UtilPrebuildBucket,
+			Instructions:       instructions,
+		})
+
```
Comment (msuozzo) [inline pkg/rebuild/rebuild/rebuildremote.go:392]: remove newline

**Review by msuozzo — APPROVED** [review]

File: pkg/rebuild/debian/strategy.go (hunk 14 above)
Comment (wbxyz) [reply pkg/rebuild/debian/strategy.go:82]: Probably?

**Pull request merged**